cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,262 @@
+
+    if (!zi || zi->ifindex == 0)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = outbound_cb;
+
+    dbg("outbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For app pod veth pair(in host ns) egress hook
+SEC("tc")
+int app_inbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr  *eth = data;
+    struct iphdr *iph = NULL;
+    struct ztunnel_info *zi = NULL;
+    __u32 inbound_cb = INBOUND_CB;
+
+    if (skb->mark == ZTUNNEL_BYPASS_MARK)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+
+    zi = get_ztunnel_info();
+
+    if (!zi || zi->ifindex == 0) {
+        dbg("inbound_to_ztunnel unable to retrieve ztunnel_info\n");
+        return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = inbound_cb;
+
+    dbg("inbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For ztunnel pod veth pair(in host ns) ingress hook
+SEC("tc")
+int ztunnel_host_ingress(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct iphdr *iph = NULL;
+    struct app_info *pi = NULL;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP )
+        return TC_ACT_OK;
+
+    skb->mark = ZTUNNEL_BYPASS_MARK;
+
+    pi = get_app_info_from_ipv4(iph->daddr);
+    if (!pi) {
+        // ip is not in ambient managed mesh
+        dbg("ztunnel_host_ingress unable to retrieve pod info: 0x%x\n", bpf_ntohl(iph->daddr));
+        return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, pi->mac_addr, ETH_ALEN);
+
+    dbg("ztunnel redirect to app(0x%x) ifidx: %u\n", bpf_ntohl(iph->daddr), pi->ifindex);
+
+    return bpf_redirect(pi->ifindex, 0);
+}
+
+// For ztunnel pod eth0 ingress hook
```

Comment:
Could you improve this comment a bit?   Where is this hook running?  zTunnel veth?